frontend/src/Components/ScaleSelector/ScaleSelector.tsx

```diff
@@ -1,25 +1,128 @@
+import { useEffect, useRef, useState } from 'preact/hooks';
 import { ScaleName, SCALES } from '../../../../shared/scales';
 import { SELECT_CHANGE_SCALE } from '../../constants';
+import sharedClasses from '../../styles.module.css';
 import { connectToWebSocket } from '../WebSocket/WebSocket';
 import classes from './ScaleSelector.module.css';
 
-export const ScaleSelector = connectToWebSocket(({ socket: { connected, setScale } }) => (
-  <select
-    disabled={!connected}
-    name="scale"
-    class={classes.select}
-    onChange={({ target }) =>
-      setScale(SCALES[(target as HTMLSelectElement).value as ScaleName].values)
+const availableScales = Object.entries(SCALES);
+
+export const ScaleSelector = connectToWebSocket(({ socket: { connected, setScale } }) => {
+  const [open, setOpen] = useState(false);
+  const [selected, setSelected] = useState(-1);
+  const selectionButtonRef = useRef<HTMLButtonElement>(null);
+  const dropdownRef = useRef<HTMLUListElement>(null);
+
+  const close = () => {
+    setOpen(false);
+    setSelected(-1);
+  };
+
+  const toggle = () => {
+    if (open) {
+      close();
+    } else {
+      setSelected(0);
+      setOpen(true);
+    }
+  };
+
+  const handleBlur = ({ relatedTarget }: FocusEvent) => {
+    if (relatedTarget !== selectionButtonRef.current && relatedTarget !== dropdownRef.current) {
+      close();
+    }
+  };
+
+  const handleKeyboardNavigation = (event: KeyboardEvent) => {
+    switch (event.key) {
+      case 'ArrowDown':
+        setSelected((selected + 1) % availableScales.length);
+        event.preventDefault();
+        break;
+      case 'ArrowUp':
+        setSelected((Math.max(selected, 0) + availableScales.length - 1) % availableScales.length);
```

Comment:
Adding of `availableScales.length` seems unnecessary to me since the modulo operator already ensures that the result is non-negative. 

Hence, I would simply suggest

```suggestion
        setSelected((Math.max(selected, 0) - 1) % availableScales.length);
```

Reply:
Unfortunately, `-1 % 2 === -1`. But we can go for a `Math.abs` instead

Reply:
> Unfortunately, `-1 % 2 === -1`. ~But we can go for a `Math.abs` instead~ Ah, but that is not the same thing...

I did not know that! I explicitly tested, but only `-2 % 2 === 0` :(